### PR TITLE
Rewrite03 test deploy

### PR DESCRIPTION
**Pull request by fgregoribattista**
## Describe Changes

The rewrite of test 03_erc20_enforcer_tests_l1 is created.

```
Suite result: ok. 76 passed; 0 failed; 8 skipped; finished in 36.08ms (67.78ms CPU time)
Ran 1 test suite in 326.71ms (36.08ms CPU time): 76 tests passed, 0 failed, 8 skipped (84 total tests)
```

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [ ] All Hardhat tests are passing

### CODE REVIEW

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+    holographERC20.supportsInterface(selector);
+  }
+
+  function testUntilDeployTransferFrom() public {
+    bytes4 selector = holographERC20.transferFrom.selector;
+    holographERC20.supportsInterface(selector);
+  }
+
+  function testUntilDeployERC20() public {
+    bytes4 computedId = bytes4(
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol:120]: suggestion: It's a very minor comment, but even thou we are checking for all the function signatures that represent a valid ERC20 according to the EIP20, I would create another test to check specifically for the interface. The reasoning behind this is that, in the future, any new function added to the contract might impact in this test being modified to add the new function, while the test for the IERC20 interface would still be checking for the standard interface id.

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+  function decreaseAllowanceToAlice(uint256 amount) public {
+    vm.prank(deployer);
+    holographERC20.decreaseAllowance(alice, amount);
+  }
+
+  /*
+   * INIT INTERFACES
+   */
+
+  function testUntilDeploySupportinterface() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: suggestion: let's remove the UntilDeploy part of the naming if not necessary.

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+    holographERC20.supportsInterface(selector);
+  }
+
+  function testUntilDeployERC20BurnInterface() public {
+    bytes4 computedId = bytes4(holographERC20.burn.selector ^ holographERC20.burnFrom.selector);
+    assertTrue(holographERC20.supportsInterface(computedId));
+  }
+
+  function testUntilDeploySafeTransfer() public {
+    bytes memory selector = abi.encodeWithSelector(bytes4(keccak256("safeTransfer(address,uint256)")));
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: suggestion: the encodeWithSelector is not needed here, just (which already is a bytes4)

```
bytes memory selector = bytes4(keccak256("safeTransfer(address,uint256)"));
```

[Here's](https://remix.ethereum.org/#code=CmNvbnRyYWN0IFN0b3JhZ2UgewogICAgdWludDI1NiBudW1iZXI7CgogICAgZnVuY3Rpb24gcmV0cmlldmUoKSBwdWJsaWMgdmlldyByZXR1cm5zICh1aW50MjU2KXsKICAgICAgICByZXR1cm4gbnVtYmVyOwogICAgfQoKICAgIGZ1bmN0aW9uIHRlc3QoKSBwdWJsaWMgcHVyZSByZXR1cm5zIChieXRlczQpewogICAgICAgIHJldHVybiB0aGlzLnJldHJpZXZlLnNlbGVjdG9yOwogICAgfSAvLyByZXR1cm5zIDB4MmU2NGNlYzEKCiAgICBmdW5jdGlvbiB0ZXN0MigpIHB1YmxpYyBwdXJlIHJldHVybnMgKGJ5dGVzNCl7CiAgICAgICAgcmV0dXJuICBieXRlczQoa2VjY2FrMjU2KGJ5dGVzKCJyZXRyaWV2ZSgpIikpKTsKICAgIH0gLy8gcmV0dXJucyAweDJlNjRjZWMxCgogICAgZnVuY3Rpb24gdGVzdDMoKSBwdWJsaWMgcHVyZSByZXR1cm5zIChieXRlczQpewogICAgICAgIHJldHVybiAgYnl0ZXM0KGFiaS5lbmNvZGVXaXRoU2lnbmF0dXJlKCJyZXRyaWV2ZSgpIikpOwogICAgfSAvLyByZXR1cm5zIDB4MmU2NGNlYzEKCiAgICBmdW5jdGlvbiB0ZXN0NCgpIHB1YmxpYyBwdXJlIHJldHVybnMgKGJ5dGVzNCl7CiAgICAgICAgcmV0dXJuICBieXRlczQoYWJpLmVuY29kZVdpdGhTZWxlY3RvcihieXRlczQoa2VjY2FrMjU2KCJyZXRyaWV2ZSgpIikpKSk7CiAgICB9IC8vIHJldHVybnMgMHgyZTY0Y2VjMQoKICAgIGZ1bmN0aW9uIHRlc3Q1KCkgcHVibGljIHB1cmUgcmV0dXJucyAoYnl0ZXM0KXsKICAgICAgICByZXR1cm4gYnl0ZXM0KGtlY2NhazI1NigicmV0cmlldmUoKSIpKTsKICAgIH0gLy8gcmV0dXJucyAweDJlNjRjZWMxIAp9) an example of different ways of obtaining the selector

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+    mintToAlice();
+    assertEq(holographERC20.balanceOf(alice), initialValue);
+  }
+
+  /*
+   * ERC20 TEST
+   *    Tokens Approvals
+   */
+
+  function testApporvalRevertZeroAddress() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: typo: `testApporvalRevertZeroAddress` => `testApprovalRevertZeroAddress`

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+    vm.expectRevert("ERC20: operator not contract");
+    holographERC20.onERC20Received(
+      deployer,
+      deployer,
+      initialValue,
+      bytes(abi.encode("0x0000000000000000000000000000000000000000"))
+    );
+  }
+
+  //TODO see why mock token have balance? remove Fail to the name of the function
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol:383]: question: is this TODO still pending? if not, let's remove it

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+  function testErc20RecivedFakeContractRevert() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: typo: `testErc20RecivedFakeContractRevert` => `testErc20ReceivedFakeContractRevert`

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 7]
```diff
@@ -0,0 +1,697 @@
...
+    holographERC20.onERC20Received(
+      address(erc20Mock),
+      deployer,
+      initialValue,
+      bytes(abi.encode("0x0000000000000000000000000000000000000000"))
+    );
+  }
+
+  //TODO see why revert ( amount exceeds balance, need mint and then not fail... ) and not non ERC20Receiver,  remove Fail to the name of the function
+  function testSafeTransferBrokenErc20RecivedRevert() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: suggestion: same here on typo and TODO

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 8]
```diff
@@ -0,0 +1,697 @@
...
+    vm.skip(true);
+    erc20Mock.toggleWorks(false);
+    vm.expectRevert("ERC20: non ERC20Receiver");
+    vm.prank(deployer);
+    holographERC20.safeTransfer(address(erc20Mock), initialValue);
+  }
+
+  //TODO see why revert ( amount exceeds balance,need mint and then not fail... ) and not non ERC20Receiver,  remove Fail to the name of the function
+  function testSafeTransferBytesBrokenErc20RecivedRevert() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: suggestion: same here on typo and TODO

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 9]
```diff
@@ -0,0 +1,697 @@
...
+    vm.prank(deployer);
+    holographERC20.safeTransfer(
+      address(erc20Mock),
+      initialValue,
+      bytes(abi.encode("0x0000000000000000000000000000000000000000"))
+    );
+  }
+
+  //TODO see why not revert,  remove Fail to the name of the function
+  function testSafeTransferFromBrokenErc20RecivedRevert() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: suggestion: same here on typo and TODO

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+
+    approvalToAlice(maxValue);
+    vm.expectRevert("ERC20: non ERC20Receiver");
+    vm.prank(alice);
+
+    holographERC20.safeTransferFrom(address(deployer), address(erc20Mock), initialValue);
+  }
+
+  //TODO see why not revert,  remove Fail to the name of the function
+  function testSafeTransferFromBytesBrokenErc20RecivedRevert() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol:432]: suggestion: same here on typo and TODO

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 11]
```diff
@@ -0,0 +1,697 @@
...
+    mintToDeployer();
+    approvalToAlice(initialValue);
+    vm.expectEmit(true, true, false, true);
+    emit Transfer(address(deployer), address(alice), initialValue);
+    vm.prank(alice);
+    holographERC20.safeTransferFrom(address(deployer), address(alice), initialValue);
+  }
+
+  function testBalanceOfDeployerAfterSafeTransferFrom() public {
+    testSafeTransferFrom();
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol:490]: praise: interesting approach!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 12]
```diff
@@ -0,0 +1,697 @@
...
+    vm.prank(alice);
+    holographERC20.safeTransferFrom(address(deployer), address(erc20Mock), initialValue);
+    //check allowance alice = 0
+    assertEq(holographERC20.allowance(deployer, alice), 0);
+  }
+
+  function testSafeTransferFromBytesToErc20() public {
+    mintToDeployer();
+    erc20Mock.toggleWorks(true);
+    vm.expectEmit(true, true, false, true);
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: question: does it make sense to check for the emission of the Approval on every call to the function approvalToAlice? I feel that these expectEmit  s before the approvalToAlice calls are not needed.

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 13]
```diff
@@ -0,0 +1,697 @@
...
+  }
+
+  /*
+   * ERC20 TEST
+   *    Permit
+   */
+
+  //TODO buildDomainSeperator
+
+  function testPermitZeroNounce() public {
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: typo: `testPermitZeroNounce` => `testPermitZeroNonce`

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 14]
```diff
@@ -0,0 +1,697 @@
...
+
+  function testPermitBadDeadLineRevert() public {
+    vm.expectRevert("ERC20: expired deadline");
+    holographERC20.permit(deployer, alice, initialValue, badDeadLine, uint8(0x00), zeroSignature, zeroSignature);
+  }
+
+  function testPermitEmptySignatureRevert() public {
+    console.log("block.timestamp", block.timestamp);
+    console.log("block.goodDeadLine", goodDeadLine);
+    console.log("block.badDeadLine", badDeadLine);
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol]: question: why do we need these console.logs? can't we remove them?

File: test/foundry/deploy/03_Erc20Enforcer.t.sol [hunk 15]
```diff
@@ -0,0 +1,697 @@
...
+    assertEq(sampleERC20.isOwner(), false);
+  }
+
+  function testErc20GetOwnerProxy() public {
+    assertEq(holographERC20.getOwner(), address(admin));
+  }
+
+  function testErc20DeployerTransferOwnerRevert() public {
+    vm.expectRevert("HOLOGRAPH: owner only function");
+    holographERC20.setOwner(alice);
```
Comment (lucas-marc) [inline test/foundry/deploy/03_Erc20Enforcer.t.sol:665]: suggestion: let's add a vm.prank(alice) here, to enforce the idea that the caller is not the owner

**Review by lucas-marc — APPROVED** [review]
Left some comments

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 2 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: Done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 14 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 5 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:383]: yes, if it is part of the skip, which is also in hardhat.
When looking for the bug it was found to be balanced, hence the comment.

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 15 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:665]: ok

Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:665]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 13 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 4 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 11 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:490]: thanks :)

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 12 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: Yes, you are absolutely right, it was already test in the previous one.

Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:546]: also the assert of allowance to alice

Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 1 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:120]: yes I find it very useful, I generate an erc20Interface test and a holographERC20 interface test. Very good suggestion! 👍🏼

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 3 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:169]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 1 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:120]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 6 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 8 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 7 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 10 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol:432]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 9 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

File: test/foundry/deploy/03_Erc20Enforcer.t.sol (hunk 3 above)
Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: yes you are right, I refactor all the safe transfer interfaces that use this type of validation 👍🏼

Comment (fgregoribattista) [reply test/foundry/deploy/03_Erc20Enforcer.t.sol]: done!

**Pull request merged**